initial contribution of a Jeelink binding

vbier opened the pull request:
This adds support for Jeelink USB sticks and connected LaCrosse Temperature sensors.

Signed-off-by: Volker Bier volker.bier@web.de

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/AbstractJeeLinkConnection.java:
```diff
@@ -0,0 +1,131 @@
+    private void initializeDevice() {
+        try {
+            if (initCommands != null) {
+                logger.debug("Initializing device on port {} with commands {} ", port, Arrays.toString(initCommands));
```
The array to String conversion will always be performed (also if you are not running in debug), you could put it in a check for logger.isDebugEnabled. I don't know how often we execute this line, most likely the impact is really minor.

vbier replied:
fixed.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkHandler.java:
```diff
@@ -0,0 +1,94 @@
+            }
+            logger.info("updating JeeLinkHandler for port {} to be ONLINE...", connection.getPort());
+            updateStatus(ThingStatus.ONLINE);
+            logger.info("JeeLinkHandler for port {} is ONLINE.", connection.getPort());
```
I believe log info should be used sparsely, since online is also visible in the UI I don't know whether you need (two) info statements for this.

vbier replied:
fixed.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkHandler.java:
```diff
@@ -0,0 +1,94 @@
+            logger.info("updating JeeLinkHandler for port {} to be ONLINE...", connection.getPort());
+            updateStatus(ThingStatus.ONLINE);
+            logger.info("JeeLinkHandler for port {} is ONLINE.", connection.getPort());
+        } catch (Exception ex) {
```
Do you really need to catch all exceptions or is only a specific subset enough?

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkReadingListener.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Volker Bier - Initial contribution
+ */
+public class JeeLinkReadingListener<R extends Reading> {
```
If it is a base class should it be abstract?

vbier replied:
It is not abstract, as it provides functionality to keep track of the current reading and can be used as is. I could try to improve the apidoc to make that clear.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkReadingListener.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Constructs a reading listener that only keeps track of the current reading.
+     */
+    public JeeLinkReadingListener() {
```
_No need for this default constructor..._

vbier replied:
It does not seem to work without (maybe due to the generics?).

martinvw replied:
I was wrong, it was one of the multiple constructors, that's why it is needed

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,231 @@
+                        // make sure status is online as soon as we get a reading
+                        updateStatus(ThingStatus.ONLINE);
+                    } catch (Throwable th) {
+                        logger.error("Uncaught throwable in JeeLink sensor handler for thing " + getThing().getLabel()
```
A self composed logging string

vbier replied:
fixed.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,231 @@
+
+                        // make sure status is online as soon as we get a reading
+                        updateStatus(ThingStatus.ONLINE);
+                    } catch (Throwable th) {
```
Don't catch throwables: http://stackoverflow.com/questions/6083248/is-it-a-bad-practice-to-catch-throwable#answer-6083271

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,231 @@
+                if (listener != null) {
+                    try {
+                        updateReadingStates(listener.getCurrentReading());
+                    } catch (Throwable th) {
```
Don't catch `Throwable` please search all

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,231 @@
+                    try {
+                        updateReadingStates(listener.getCurrentReading());
+                    } catch (Throwable th) {
+                        logger.error("Uncaught throwable in JeeLink sensor handler for thing " + getThing().getLabel()
```
Self composed logging string, can you check your branch with the check report which has been integrated in the build, after a `mvn clean install` your `target` directory should contain a `code-analysis/report.html`

vbier replied:
I have fixed this occurance, but I do not have a code-analysis directory after doing a mvn clean install. Is there anything else i have to do in order to get this?

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,231 @@
+        };
+        logger.debug("JeeLink sensor timeout job for thing {} ({}) created with interval {} s", getThing().getLabel(),
+                getThing().getUID(), sensorTimeout);
+        return scheduler.scheduleAtFixedRate(runnable, sensorTimeout, 1, TimeUnit.SECONDS);
```
Feels quite often what will it do? For how many nodes? What is the load for this on a device with a small footprint?

martinvw replied:
The coding guidelines state the following:
> For periodically executed jobs that do not require a fixed rate scheduleWithFixedDelay should be preferred over scheduleAtFixedRate.

vbier replied:
Changed to scheduleWithFixedDelay. As you can see, it only checks the status, computes a difference and does a comparison. This should be as light-weight as it gets.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSerialConnection.java:
```diff
@@ -0,0 +1,112 @@
+                                    propagateLine(input.readLine());
+                                }
+                            } catch (IOException ex) {
+                                logger.error("IOException reading from port {}!", port);
```
Should we log the exception here, or can it only have a single cause?

vbier replied:
fixed.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSerialConnection.java:
```diff
@@ -0,0 +1,112 @@
+                    serialPort.notifyOnDataAvailable(true);
+                } catch (UnsupportedCommOperationException | IOException | TooManyListenersException ex) {
+                    closeConnection();
+                    throw ex;
```
You could also wrap it here and let it trough in the outer catch, now you catch these exceptions twice. Should the method maybe be split? A single method contains now three layers of nested try-catch blocks.

vbier replied:
I have remove the middle catch.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,125 @@
+ * @author Volker Bier - Initial contribution
+ */
+public class JeeLinkTcpConnection extends AbstractJeeLinkConnection {
+    private static final Pattern IP_P = Pattern.compile("([0-9]+\\.[0-9]+\\.[0-9]+\\.[0-9]+):([0-9]+)");
```
`IP_P` could maybe use a clearer name

vbier replied:
fixed.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,125 @@
+            reader = new Reader(socket);
+            reader.start();
+        } catch (IOException ex) {
+            logger.error("Failed to create socket.", ex);
```
do you need to close any sockets or readers in this case?

vbier replied:
fixed.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,125 @@
+            socket = new Socket(hostName, portNumber);
+            logger.debug("TCP socket created.");
+            reader = new Reader(socket);
+            reader.start();
```
Coding guidelines:
> Creation of threads must be avoided. Instead, resort into using existing schedulers which use pre-configured thread pools. If there is no suitable scheduler available, start a discussion in the forum about it rather than creating a thread by yourself.

vbier replied:
I took that from the DSCAlarm binding, which does it exactly in the same way (I know this is a weak argument :wink:). 

I am not sure if using a scheduler makes much sense for a thread that lives from the time the thing is connected until thing removal/openhab shutdown.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/ec3k/Ec3kReadingConverter.java:
```diff
@@ -0,0 +1,80 @@
+            Matcher matcher = LINE_P.matcher(inputLine);
+            if (matcher.matches()) {
+                /*
+                 * [ec3kSerial.1]
```
`[ec3kSerial.1]` where does that refer to? Is it a protocol, a piece of code from someone else?

vbier replied:
Does not really belong here. Is part of the header that is written by the sketch running on the USB stick.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/lacrosse/LaCrosseTemperatureReadingConverter.java:
```diff
@@ -0,0 +1,76 @@
+        if (listenerIdMap.size() > 0 && inputLine != null) {
+            Matcher matcher = LINE_P.matcher(inputLine);
+            if (matcher.matches()) {
+                // Format
```
Nice looks clear!

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/JeeLinkHandlerFactory.java:
```diff
@@ -0,0 +1,88 @@
+            final String sketchName = (String) thing.getConfiguration().getProperties().get(PROPERTY_SKETCH_NAME);
+            logger.info("creating JeeLinkHandler with sketch {} for thing {}...", sketchName, thing.getUID().getId());
+
+            JeeLinkHandler jlh = new JeeLinkHandler((Bridge) thing);
```
_ihmo you don't need the jlh field, might be a matter of preference_

vbier replied:
fixed.

vbier commented on addons/binding/org.openhab.binding.jeelink/pom.xml:
```diff
@@ -0,0 +1,30 @@
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
```
fixed.

vbier commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/JeeLinkBindingConstants.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Copyright (c) 2014-2016 by the respective copyright holders.
```
fixed.

vbier commented on addons/binding/org.openhab.binding.jeelink/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,26 @@
+Bundle-Name: JeeLink Binding
+Bundle-SymbolicName: org.openhab.binding.jeelink;singleton:=true
+Bundle-Vendor: openHAB
+Bundle-Version: 2.0.0.qualifier
```
fixed.

vbier commented on addons/binding/org.openhab.binding.jeelink/build.properties:
```diff
@@ -0,0 +1,6 @@
+bin.includes = META-INF/,\
+               .,\
+               OSGI-INF/,\
+               ESH-INF/
```
fixed. not really sure why this is still shown here?

martinvw commented:
You still seem to have code analysis warnings:

[ERROR] ERROR found by checkstyle: .binding.jeelink/about.html:0 Missing about.html file.
[ERROR] ERROR found by findbugs: org.openhab.binding.jeelink.handler.JeeLinkTcpConnection$Reader.java:124 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance
[ERROR] ERROR found by findbugs: org.openhab.binding.jeelink.handler.JeeLinkTcpConnection$Reader.java:137 Format should be constant. Use placeholder to reduce the needless cost of parameter construction. see http://www.slf4j.org/faq.html#logging_performance

Note you can also run it locally and view the complete report.

martinvw commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,141 @@
+
+        try {
+            reader = new Reader(socket);
+            reader.start();
```
The main reason to use on of those scheduler is that we have fixed amount of thread to run the background tasks, this will help keeping the resource usage under control. So its more about the thread pool then the scheduling.

martinvw replied:
`scheduler.execute` might be the appropriate entry, but others might know better.

vbier commented:
It looks like I am missing something.

I have executed mvn clean install -P test (as otherwise it does not do static code analysis). This does not give me any errors, but I see log messages like 
`[INFO] Adding dependency to checkstyle:7.2
[WARNING] Unable to locate Source XRef to link to - DISABLED
[INFO] 
[INFO] --- static-code-analysis:0.0.2:pmd (default) @ org.openhab.binding.allplay ---
[WARNING] Unable to locate Source XRef to link to - DISABLED
[WARNING] Unable to locate Source XRef to link to - DISABLED
`
It does not make a difference if I execute maven from the addon dir or from the toplevel dir.

martinvw commented:
You are using both an old version and an unsupported maven version, it has been fixed in the next version, so you either rebase all your changes, requires some git-kung-fu which can lead to losing all your code, or just let travis test it for you :-)

martinvw commented:
@vbier can you please do some testing with the scheduler execute?

vbier commented:
Okay, I have changed this to use the scheduler. Looks like this works, I event found a bug in the TCP code (which I normally do not use).

martinvw commented:
@vbier some guidelines changed (see also https://github.com/openhab/openhab2-addons/pull/2179) can you please add the `about.html` to the `build.properties. Thanks!

vbier commented:
I have added the about.html file to the binary build.

martinvw commented:
Thanks!

martinvw commented:
Triggering a rebuild

overrunner commented:
Maybe i found an issue.
I connect my jeelink via ser2net on my openhab instance.
tcp://192.168.0.8:3000

When i restart the jeelink ser2net host (a raspi), the Jeelink gets status "ONLINE" in openhab, but every sensor stays OFFLINE. Even if i change the IP adress to nonsense and insert the correct one.
Only after restarting the openhab2 service, the sensors are found again.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/ESH-INF/i18n/jeelink_de.properties:
```diff
@@ -0,0 +1,59 @@
+channel-type.sensor-time-channel.label = Einschaltzeit des Sensoren
+channel-type.sensor-time-channel.description = Die Anzahl der Sekunden w\u00e4hrend deren der Sensor eingeschaltet war.
+channel-type.resets-channel.label = Resets
+channel-type.resets-channel.description = ???
```
You do not know it yourself? Then maybe simply leave the description empty or remove the channel altogether.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+        <description>@text/thing-type.ec3k.description</description>
+
+		<channels>
+			<channel id="sensorId" typeId="sensor-id-channel" />
```
This looks dodgy - the sensorId is already a config parameter of the Thing itself, so why should here be a channel for it? It won't ever change, will it?

vbier replied:
It changes if the battery in the sensor is changed. Since I am not really that knowledgeable with respect to openhab concepts and api, I had added a channel with a dedicated command to set the id. I will remove this completely, as now the id can be changed in the properties.

kaikreuzer replied:
You mean every time that a battery is replaced, the sensor gets a new id? How is a sensor identified apart from its id (which does not seem to fulfill this purpose then...)?

vbier replied:
There is no other identification, its the users responsibility to update the id once he changes batteries.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+	</channel-type>
+
+	<!-- Resets Channel Type -->
+	<channel-type id="resets-channel">
```
If you really think this is needed (what does it do?), have you at least considered to mark it "advanced=true"?

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+	</channel-type>
+
+	<!-- Sensor Type Channel Type -->
+	<channel-type id="sensor-type-channel">
```
This does not seem to be a useful channel - similar to the sensorId it will never change. So if this value is determined dynamically, please set it as a Thing property instead.

vbier replied:
This is part of the protocol and is reported by the sensor with either 1 or 2 as a value. I have no idea what it means but thought it would be better to make it available for people that did. I will remove it.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/discovery/SensorDiscoveryService.java:
```diff
@@ -0,0 +1,139 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.jeelink.discovery;
```
You might want to move that to an internal package, e.g. `org.openhab.binding.jeelink.internal.discovery`.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/discovery/SensorDiscoveryService.java:
```diff
@@ -0,0 +1,139 @@
+                    ThingUID sensorThing = new ThingUID(def.getThingTypeUID(), newId);
+
+                    DiscoveryResult discoveryResult = DiscoveryResultBuilder.create(sensorThing)
+                            .withLabel(def.getName() + " (ID = " + id + ")").withBridge(bridge.getThing().getUID())
```
As you set the sensor Id as a property as well, you should remove it from the label, but rather use `withRepresentationProperty("id")`, which indicated to the UI that this should be shown for identification.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/AbstractJeeLinkConnection.java:
```diff
@@ -0,0 +1,137 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.jeelink.handler;
```
You should consider moving this (and many other classes) to an internal package like `org.openhab.binding.jeelink.internal`.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/AbstractJeeLinkConnection.java:
```diff
@@ -0,0 +1,137 @@
+                        w.write(cmd);
+                    }
+                } else {
+                    logger.warn("Connection on port {} did not provide an init stream for writing init commands", port);
```
Don't log warnings, but set the Thing status to OFFLINE with an according error message.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/AbstractJeeLinkConnection.java:
```diff
@@ -0,0 +1,137 @@
+                }
+            }
+        } catch (IOException e) {
+            logger.error("Failed to send init commands to port {}: {}", port, e.getMessage());
```
Don't log errors, but set the Thing status to OFFLINE with an according error message.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkConnection.java:
```diff
@@ -0,0 +1,37 @@
+import java.io.OutputStream;
+
+/**
+ * Interface for connections to JeeLink UDB Receivers.
```
USB?

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkHandler.java:
```diff
@@ -0,0 +1,116 @@
+
+    @Override
+    public void handleCommand(ChannelUID channelUid, Command command) {
+        logger.error("Thing {} channel {} does not support sending of commands!", getThing().getUID(),
```
Don't log any errors here, but simply ignore commands that you cannot handle.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,221 @@
+ * @author Volker Bier - Initial contribution
+ */
+public abstract class JeeLinkSensorHandler<R extends Reading> extends BaseThingHandler {
+    private Logger logger = LoggerFactory.getLogger(JeeLinkSensorHandler.class);
```
add final

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,221 @@
+                logger.warn("Received invalid command {}!", cmdName);
+            }
+        } else if (!(command instanceof RefreshType)) {
+            logger.warn("Thing {} ({}) channel {} does not support sending of commands!", getThing().getLabel(),
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,221 @@
+
+            Matcher matcher = Pattern.compile("SetSensorId (\\w+)").matcher(cmdName);
+            if (matcher.matches()) {
+                final String oldId = getThing().getProperties().get(PROPERTY_SENSOR_ID);
```
What is going on here? Channels definitely must not be used to set Thing properties... What's the use case here?

vbier replied:
see above. will be removed.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,221 @@
+            c.removeReadingListener(listener);
+            c.addReadingListener(newId, listener);
+        }
+        logger.info("Thing {} ({}) updated sensor id from {} to {}", getThing().getLabel(), getThing().getUID(), oldId,
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,221 @@
+    @Override
+    public synchronized void initialize() {
+        if (initialized) {
+            logger.info("JeeLink sensor handler for thing {} ({}) is already initialized", getThing().getLabel(),
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,221 @@
+        } catch (NumberFormatException ex) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                    "Thing has an invalid sensor id: " + idStr);
+            logger.info("JeeLink sensor handler for thing {} ({}) is OFFLINE", getThing().getLabel(),
```
reduce to debug

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSensorHandler.java:
```diff
@@ -0,0 +1,221 @@
+        };
+        logger.debug("JeeLink sensor handler value propagation job for thing {} ({}) created with interval {} s",
+                getThing().getLabel(), getThing().getUID(), updateInterval);
+        return scheduler.scheduleAtFixedRate(runnable, updateInterval, updateInterval, TimeUnit.SECONDS);
```
Would `withFixedDelay` also work here? This would definitely be preferred.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkSerialConnection.java:
```diff
@@ -0,0 +1,115 @@
+                                propagateLine(input.readLine());
+                            }
+                        } catch (IOException ex) {
+                            logger.error("IOException reading from port {}!", port, ex);
```
don't log an error, but set the Thing status accordingly

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,152 @@
+            socket.setKeepAlive(true);
+            logger.debug("TCP socket created.");
+        } catch (IOException ex) {
+            logger.error("Failed to create socket.", ex);
```
don't log an error, but set the Thing status accordingly.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,152 @@
+            scheduler.execute(reader);
+        } catch (IOException ex) {
+            closeSocketSilently();
+            logger.error("Failed to create reader.", ex);
```
don't log an error

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,152 @@
+                }
+            } catch (IOException ex) {
+                if (isRunning) {
+                    logger.error("Error reading from TCP port {}!", port, ex);
```
don't log an error

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/handler/JeeLinkTcpConnection.java:
```diff
@@ -0,0 +1,152 @@
+                socket.close();
+                inputReader.close();
+            } catch (IOException ex) {
+                logger.error("Failed to close TCP port {}!", port, ex);
```
don't log an error

vbier commented on addons/binding/org.openhab.binding.jeelink/README.md:
```diff
@@ -0,0 +1,135 @@
+| Sketch Name       | String       | Currently only LaCrosseITPlusReader is supported
+| Init Commands     | String       | A semicolon separated list of init commands that will be send to the Jeelink, e.g. "0a v" for disabling the LED.
+
+The Port Name has to be of the form:
```
I have no idea how an items file looks like. All my items are created with the paper UI.

vbier replied:
Logically it does not make sense to me to have a different thing for the same piece of hardware. What if I change this to support either normal serial ports (without prefix) and additionally allow to specify hostname and port by using the tcp:// prefix?

kaikreuzer replied:
This won't work once we want to offer a dropdown list of available (serial) ports.

How is it the same hardware if it is connected either through USB or available through IP on the network? Why shouldn't this be treated as different types? Note that all it requires is the definition of an additional bridge-type - both types can then use the very same handler implementation.

vbier replied:
People connect the usb stick to another device on their network in order to improve reception. They then use ser2net to provide the usb port over tcp to the openhab server.

kaikreuzer replied:
ok, thanks for the explanation. Nonetheless, I think that in that case it is not the stick that you talk to, but the remote host that you connect to; other bindings also have different bridges for TCP or serial, see e.g. https://github.com/openhab/openhab2-addons/tree/master/addons/binding/org.openhab.binding.dscalarm/ESH-INF/thing.

vbier commented on addons/binding/org.openhab.binding.jeelink/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,224 @@
+			<channel id="sensorType" typeId="sensor-type-channel" />
+			<channel id="temperature" typeId="temperature-channel" />
+			<channel id="humidity" typeId="humidity-channel" />
+			<channel id="batteryNew" typeId="battery-new-channel" />
```
Since battery new and low are only boolean flags i am not sure if this will result in any meaningful value for a level type.

kaikreuzer replied:
Well, you could cover 100%, 66% and 33% values, which is as good as it gets with that information. It is at least as meaningful as the two boolean flags...

vbier replied:
But the percentages would not be credible at all and people will be surprised as they do not know that this is only a fake value. A few days after you insert new batteries you are at 66%, two years later you are at 33% and a week later the batteries are dead.

kaikreuzer replied:
It is not fake, it is an indication. The percent values are meant to provide a battery level for showing an according icon (which usually anyhow only has ~4 steps). There isn't a single battery powered device where such values are anywhere close to be precise.

vbier commented on addons/binding/org.openhab.binding.jeelink/README.md:
```diff
@@ -0,0 +1,135 @@
+|-------------------------|-----------------|----------------------|------------
+| sensorId                | String          | SetSensorId _newID_  | Sets the sensor ID of the current sensor to the specified _newID_
+
+
```
I have no idea how a things file looks like. All my things are created with the paper UI.

kaikreuzer replied:
You will find MANY examples on all the other binding READMEs :-)

vbier commented:
I have fixed most of the smaller things that have been mentioned by you, @kaikreuzer. Open is the change with the serial port and a refactoring to move stuff to internal packages. I will address these problems later. Is there any documentation on what should go into internal packages?

Another thing I found out while making the changes to update the thing status with an error message instead of logging is that the serial library seems to have problems with connection loss/reconnect. 
When I unplug the USB stick, I only sometimes get an exception, most of the time it just seems to hang. If I plug the USB stick back in, OH2 restarts automatically (when an exception had been reported earlier), or simply does nothing. 
I have tested the reconnect with ser2net, where it works without problems, so this does not seem to be a problem in the binding itself, but rather in the library.

kaikreuzer commented:
> Is there any documentation on what should go into internal packages?

No, but in general, EVERYTHING should go in an internal package, which is not meant to be shared with any other bundle.

> I will address these problems later.

So shall I wait or would you say that we are in a state that is ready to be merged?

vbier commented:
I would prefer to make the serial port changes before the binding is merged. Once it is part of the official distribution there should not directly be a breaking change that forces user to recreate the things. The refactoring can be done after the merge.

kaikreuzer commented:
ok, so please ping me once you are done!

vbier commented:
I have added a dedicated bridge type for network connected sticks and additionally squashed the commits and updated to version 2.2.0-SNAPSHOT.

I also have an uncommited refactoring that puts most classes into internal packages and also separates different functionality better into different classes. I am not sure if I should submit that now (or in a separate pull request later), as it changes huge parts of the code that would then have to be reviewed again.

vbier commented:
I had the refactored code running on my pc for about a month without problems, so I have now pushed the refactoring. I also added a thing configuration example to the README.

kaikreuzer commented:
Sounds as if I should put it high on my to-do list then - will try to find the time later this week! Thanks for your patience!

vbier commented:
I have patiene with you as long as you have patience with me. I just had to install openhab from scratch and realized that the discovery no longer works. I will look into this and tell you when it is fixed.

vbier commented:
Forget my last comment. I am just too stupid to use my own binding. I removed discovery for the stick, so it is expected that none is found. After creating it manually, all sensors are discovered as expected.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/connection/JeeLinkConnection.java:
```diff
@@ -0,0 +1,29 @@
+ * @author Volker Bier - Initial contribution
+ */
+public interface JeeLinkConnection {
+    void closeConnection();
```
Would be nice to have some JavaDoc on interface methods.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/ec3k/Ec3kSensorHandler.java:
```diff
@@ -0,0 +1,87 @@
+                            getThing().getUID().getId(), currentWatt, reading.getCurrentWatt(), maxWatt,
+                            reading.getConsumptionTotal(), reading.getApplianceTime(), reading.getSensorTime());
+
+                    updateState(getThing().getChannel(CURRENT_WATT_CHANNEL).getUID(), new DecimalType(currentWatt));
```
You can use the id instead of the uid here:
```
updateState(CURRENT_WATT_CHANNEL), new DecimalType(currentWatt));
```

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/lacrosse/LaCrosseTemperatureSensorHandler.java:
```diff
@@ -0,0 +1,90 @@
+                            "updating states for thing {} ({}): temp={} ({}), humidity={}, batteryNew={}, batteryLow={}",
+                            getThing().getLabel(), getThing().getUID().getId(), temp, reading.getTemperature(),
+                            reading.getHumidity(), reading.isBatteryNew(), reading.isBatteryLow());
+                    updateState(getThing().getChannel(TEMPERATURE_CHANNEL).getUID(), new DecimalType(temp));
```
use id instead of uid

vbier commented on addons/binding/org.openhab.binding.jeelink/ESH-INF/i18n/jeelink_de.properties:
```diff
@@ -0,0 +1,63 @@
+# binding
+binding.jeelink.name = JeeLink Binding
+binding.jeelink.description = Binding f\u00fcr JeeLink USB Empf\u00e4nger und EC3000 Energiekosten-Messger\u00e4te / Lacrosse Temperatur-Sensoren.
```
I had Umlauts in there a while ago, but that did not work at the time. Has this changed or did I do something else wrong? I have changed this now to using Umlauts and it works.

vbier commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/JeeLinkHandler.java:
```diff
@@ -0,0 +1,161 @@
+
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, cause);
+
+        connectJob = scheduler.schedule(connector, 10, TimeUnit.SECONDS);
```
That's not fair. Java 8 was not even released when I created the pull request! ;-)

vbier commented on addons/binding/org.openhab.binding.jeelink/src/main/java/org/openhab/binding/jeelink/internal/SensorDefinition.java:
```diff
@@ -0,0 +1,112 @@
+            for (SensorDefinition<?> sensor : SENSOR_DEFS) {
+                converters.add(sensor.createConverter());
+            }
+
```
I am not sure what you mean. Do you mean the empty lines before the return statements? I do not want to remove them, as IMHO it makes the core more readable.

kaikreuzer replied:
I rather meant the ones around the "for" loop and the "if". You don't have those at [other places](https://github.com/openhab/openhab2-addons/pull/1541/files/360d439a43a7782f66e086cb34d61bd2b1a97744#diff-86b616acb370dce02fac3eff48e6d6d2R103) either.

vbier commented:
@kaikreuzer , is there still anything missing?

kaikreuzer commented:
Just two things:
- [x] A ping from your side that you are done with the changes.
- [ ] Some of my time to check and approve the updates.

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/README.md:
```diff
@@ -0,0 +1,117 @@
+
+## Full Example
+
+A typical Thing configuration for ec3k looks like this:
```
please add an empty line before the code section

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/README.md:
````diff
@@ -0,0 +1,117 @@
+Thing jeelink:ec3k:14E7 "ec3k 2" (jeelink:jeelinkUsb:ec3k)  @ "home" [ sensorId="14E7"]
+```
+
+A typical Thing configuration for lacrosse looks like this:
````
please add an empty line before the code section

kaikreuzer commented on addons/binding/org.openhab.binding.jeelink/README.md:
````diff
@@ -0,0 +1,117 @@
+Thing jeelink:lacrosse:sensor2 "Jeelink lacrosse 2" (jeelink:jeelinkUsb:lacrosse)  @ "home" [ sensorId="18", minTemp=10, maxTemp=32]
+```
+
+A typical item configuration for a LaCrosse temperature sensor looks like this:
````
please add an empty line before the code section

vbier commented:
There you go. Empty lines have been added.
